tutorials/haskell/generics/tutorial.md

```diff
@@ -0,0 +1,584 @@
+be noted that generics is not something academic and useless, quite the
+contrary, it's a very pragmatic way to reduce the amount of boilerplate (and
+associated with it errors) in your functional code with minimal mental
+overhead. In fact, by the time you get to the end of the tutorial, I hope
+you'll agree that generics are easy to use and easy “to get” as well.
+
+Let's start from the beginning. What is generics? **Generics**, like in
+other languages, **is a way to use the same code to manipulate different
+data types**. We should note that in this regard the technique is very close
+to the much more common notion of polymorphism, which in Haskell comes in
+two flavors:
+
+* **Parametric polymorphism**, when we have type variables in functions/data
+  types. This allows the same function, such as `const` to work with any
+  types of arguments, as long as the more general types from a function's
+  signature can be made equal to concrete types we want to work with by
+  instantiating/fixing the type variables.
+
+* **Ad-hoc polymorphism**, which allows to perform a computation abstracted
+  over instances of one or more type classes. We request that a type has
+  some properties of interest and then describe the computation in terms of
+  these properties. The code is then applicable to any data type that has
+  the properties.
+
+How are generics different? Generics allow us to describe transformation
+that works in terms of not concrete types, but general combinators that
+describe shape and meta information about a data type. This way, we can
+declare how to perform a computation on almost any data type!
+
+## The idea about generics
+
+Haskell's mechanism that opens the possibility for such a thing as generics
+is, of course, type classes and ad-hoc polymorphism. The ability to describe
+a data type in terms of a set of combinators is our property, encapsulated
+by the `Generic` type class. Data type that is an instance of that type
+class can be processed by functions that work with representation of data
+type, not the data type itself. These functions, however, are by definition
+polymorphic and also usually hidden behind a type class interface, so the
+whole thing usually takes the form of automatic deriving of a type class
+instance.
+
+The process can be described like this:
+
+1. Given data type gets `Generic` instance automatically, as it can be
+   generated by the compiler with the help of `DeriveGeneric` language
+   extension.
+
+2. There is a type class instance defined for instances of `Generic` (or
+   `Generic1`). Since our data type is now an instance of those type
+   classes, that generics-based instance starts to work for us and we
+   magically get all the useful methods of that type class.
+
+If you know about `Data` and `Typeable` type classes, then you probably know
+that it's possible to do something similar using information which methods
```

Comment:
... using the information those type classes provide.  Data and Typable are beyond the scope of this tutorial, but you can find more about them here and here. ... (We should provide links to those libraries.)